enum.c

```diff
@@ -1385,6 +1392,181 @@ sort_by_cmp(const void *ap, const void *bp, void *data)
     return OPTIMIZED_CMP(a, b);
 }
 
+
+/*
+    This is parts of uniform sort
+*/
+
+#define uless rb_uniform_is_less
+
+static inline bool
+rb_uniform_is_less(VALUE a, VALUE b)
+{
+
+    if (FIXNUM_P(a) && FIXNUM_P(b)) {
+        return (SIGNED_VALUE)a < (SIGNED_VALUE)b;
+    }
+    else if (RB_FLOAT_TYPE_P(a) && RB_FLOAT_TYPE_P(b)) {
+        return rb_float_cmp(a, b) == -1;
+    }
+    else if (FIXNUM_P(a) && RB_FLOAT_TYPE_P(b)) {
+        return rb_float_cmp(b, a) == 1;
+    }
+    else if (RB_FLOAT_TYPE_P(a) && FIXNUM_P(b)) {
+        return rb_float_cmp(a, b) == -1;
+    }
+    else {
+        return rb_cmpint(rb_funcallv(a, id_cmp, 1, &b), a, b) == -1;
+    }
```

Comment:
As both of `a` and `b` should be `Fxinum` or `Float` here, the last fallback to `rb_funcallv` should never be used.
Moreover, it must not be called without a callcc check inside `RARRAY_PTR_USE`.

```suggestion
    if (FIXNUM_P(a) && FIXNUM_P(b)) {
        return (SIGNED_VALUE)a < (SIGNED_VALUE)b;
    }
    else if (FIXNUM_P(a)) {
        RUBY_ASSERT(RB_FLOAT_TYPE_P(b));
        return rb_float_cmp(b, a) > 0;
    }
    else {
        RUBY_ASSERT(RB_FLOAT_TYPE_P(a));
        return rb_float_cmp(a, b) < 0;
    }
```